R/inject-code.R

```diff
@@ -0,0 +1,56 @@
+#' Injects code into an existing function.
+#'
+#' The `code` argument is recorded without evaluation and it will be inserted
+#' at the beginning of the `fun`. If `where` is not `onentry` it will be
+#' wrapped in `on.exit` call. If `where` is `onerror`, the wrapped code will
+#' further make the `code` to run only in the case the function failed with an
+#' error condition.
+#'
+#' @param fun a function into which the code should be inserted.
+#' @param code an expression that should be inserted.
+#' @param where a indicates at which point of the `fun` invocation the given
+#'     `code` shall run.
+#'
+#' @export
+inject_code <- function(fun, code, where=c("onentry", "onexit", "onerror")) {
```

Comment:
Consider making the API simpler in the following way -
- Rename `inject_code` to `inject`.
- Make `inject` an S3 method. I can see it being useful for injecting code in a code block or at some index in an `expression` once we start annotating call sites and conditional branches. Making it S3 will make it easier for us to extend the API. As of now, since we understand the semantics of injection in a function, we will only support `inject.function`
- It should also have an argument called `wrap` which can be TRUE or FALSE. `wrap` will wrap the code in a function body and call the function in place. One can use wrapping to prevent name collision. By default, `wrap` should be TRUE.
- Consider renaming `onentry`, `onexit` and `onerror` to `begin`, `return`, `error` respectively. The `on` prefix is an implementation detail. The API user should be exposed to an abstract API. The current naming scheme is a bit confusing, what is the difference between exit and error? Don't we exit when we error?

Reply:
> S3 method instead
Good idea. We can have `inject.function`, `inject.expression` and `inject.default` which will convert its argument into expression using `as.expression`.

> wrap
agree, although I would not make it default.

> renaming
the _on_ prefix was suppose to hint when will the execution of the code take place. The difference between `onexit` and `onerror` is that the former will run regardless on the result of the function invocation while the latter only of there was an error. I could add `onsuccess` as a counterpart of the `onerror`.

>  The API user should be exposed to an abstract API.
Not sure what do you mean

Reply:
I like `onsuccess` and `onerror` compared to onexit and onerror because `onsuccess` and `onerror` are orthogonal to each other.

Reply:
There will be always `onerror` as well in the case on wants to run the code regardless on the status